### PR TITLE
Using NameOf when throwing exceptions

### PR DESCRIPTION
**Pull request by raffaeler**
Created a new pull request as [requested here](https://github.com/dotnet/winforms/pull/115#issuecomment-444223408)

This pull request is a refactoring on all the argument exceptions specifying the name of the argument as hardcoded strings. All of them have been replaced with nameof(argumentname).
Some argument name were wrong, now they are fixed.

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 1]
```diff
@@ -419,11 +419,28 @@ public virtual void SetAudio(byte[] audioBytes)
         {
             if (audioBytes == null)
             {
-                throw new ArgumentNullException("audioBytes");
+                throw new ArgumentNullException(nameof(audioBytes));
             }
             SetAudio(new MemoryStream(audioBytes));
         }
 
+        /// <include file='doc\DataObject.uex' path='docs/doc[@for="DataObject.SetAudio"]/*' />
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: Hi @raffaeler, 
Thanks for the fixes you made! 
Can you please exclude from this PR the changes which are also included in PR#148 (https://github.com/dotnet/winforms/pull/148/files)?

Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: @JuditRose Just sent a new commit for this. Let me know

**Comment by Tanya-Solyanik** [comment]
@raffaeler - did you run a Roslyn analyzer? Is it publicly available? It's difficult to review change of this size, but might be easier to review the analyzer code.

**Comment by raffaeler** [comment]
@Tanya-Solyanik I wrote in the past a similar analyzer but I decided not run it because there are grayed regions and comments that are not covered from the changes.
So I changed strategy, created a regex and verified the replacements one by one.
Among all the changes, I did not replace a bunch of those because, for example, the string was referring to "DividerWidth" but the argument was "DividerThickness".

**Comment by raffaeler** [comment]
@MarcoRossignoli a gist for the regex? I u ndersdtand they need to verify the results one by one in any case.
@Tanya-Solyanik Sorry if the PR is large, but creating multiple PRs does not make that easier.

**Comment by MarcoRossignoli** [comment]
> I undersdtand they need to verify the results one by one in any case.

I agree but maybe with gist they can verify code and re-apply amends and run diffs to understand if there are issues. There are also amends out of scope of PR https://github.com/dotnet/winforms/pull/151/files?utf8=%E2%9C%93&diff=split#diff-d59c3102cf61d3be2dd4889d505f16d9R1054 maybe could be better submit on other PRs. But I come from coreX repos and I don't know if the rules here are the same, let us know @Tanya-Solyanik

**Comment by KlausLoeffelmann** [comment]
To be honest, I have a strong gut feeling that goes into the direction of @Tanya-Solyanik's question: I'd rather prefer an Analyzer for the code, and for the comments the RegEx would still work.

**Comment by raffaeler** [comment]
@KlausLoeffelmann given that the original source code contained wrong hardcoded strings, a manual revision would be needed in any case. In addition to that, there are cases (the ones I cited in my comment above) where the hardcoded string does not match because it refers to a width while the parameter is a thickness.
I know it is boring, but the list of replcaments is not that long.

**Comment by raffaeler** [comment]
@MarcoRossignoli changes other than the replacements are probably on a single file from a previous PR that also went closed accidentally. This PR is intended only for nameof replacements and nothing more. The fact I had to re-fork and redo the push request are probably the souce of the additional changes (which should affect one or two files, no more).

**Comment by KlausLoeffelmann** [comment]
@raffaeler Changes would result in a compiler error one way or the other, but still an Analyzer would do this semantically not syntactically.

Said that, an Analyzer could already check the correctness/compilability of the resulting CodeFix, and probably add e.g. `ObsoleteAttributes` at those spots instead of the intended CodeFix, where compilation would fail if the fix _was_ applied for wrongly hardcoded strings.

So, this would result in 3 PRs, which less Risk, since after each Analyzer pass the solution would still compile:

a) Things the Analyzer could safely do.
b) Spots where the Analyzer left ObsoleteAttributes, and which needed to be resolved manually.
c) Comments and such via RegEx. (And maybe even this work could be done by the Analyzer?).

The more I think of this, the more I tend to an Analyzer. But I am curious what others of our team think?

**Comment by raffaeler** [comment]
@KlausLoeffelmann The syntax is validated anyway by compiling, in this case the analyzer is not that different.
I wrote and use a lot of Roslyn based code, but more the analyzer is complex, more difficult is validating the analyzer code, blindly trusting the results.
That said, I do not agree on the ObsoleteAttribute and will let the team take the decision they want.

**Comment by KlausLoeffelmann** [comment]
@raffaeler : My point is that with a RegEx pass, you could not test if the resulting code for each occurance would still compile, and you would necessarily _need_ a manual cleanup before you'd able to check in. 'Obsolete' came to mind the second I wrote my reply, but you're right, it's not at all necessary: The Analyzer should just not apply any code fix in those cases, since it can find those spots for manual fixing, anyway.

My point is: I'd like to have a compilable solution at any time, and a RegEx pass simply does not guarantee that. My 2 cents.

**Comment by raffaeler** [comment]
@KlausLoeffelmann compilation is *always* guaranteed by CI checks which were enforced as the repo went public. In addition to that I always compile and run the tests before a PR, and the same can be done by the team before approving the PR when the CI is not in place.

The only thing that can be wrong is that the argument inside nameof does not match with the intent of the coder and refers to the wrong argument (that still exists since the code compiles).
In that case a manual review is needed, regardless using an analyzer, a regex or whatever else.

**Comment by KlausLoeffelmann** [comment]
Again, the main argument I have is, that if you applied the Analyzer, if should detect the cases, where the resulting fix would NOT compile, and leave it alone. That is certainly possible, right? So, that resulted in an intermediate result, which would not needed to be touched manually, would have a lot of spots fixed all ready, and could be tested and could go in. In my opinion, that would minimize the risk, because we'd break the big change into somewhat smaller changes (manually after automatically) which were much easier to review.

**Comment by raffaeler** [comment]
I understand what you are saying, but it doesn't change anything. The compile check is done in both cases. 
We can discuss about the reliability of the code written on the analyzer that can be wrong as well even if the result compiles but at the very end comparing the two listing is the only way.
Again, the list of changes is not that long.

**Comment by KlausLoeffelmann** [comment]
How would I recognize with RegEx, what came from the RegEx pass and what would be manually edited?

**Comment by KlausLoeffelmann** [comment]
Sure, I will sit with my fellow team members as soon as they come online (I am working on the WinForms team remotely from Germany, hence already awake :-) ), and discuss this.

Still, to make sure, I understood correctly:
What would - in your example - the original code line be?

**Comment by KlausLoeffelmann** [comment]
I am sorry, you included the original line.

**Comment by raffaeler** [comment]
@KlausLoeffelmann in this case, just reject the PR by yourself ...

**Comment by raffaeler** [comment]
In any case, analyzer or regex, after the replacement you just know the 'startIndex' argument is a valid token because the project compiles.
Since both the analyzer and the regex do the replacement starting from the string "startIndex", the initial intent of the developer who wrote the code is always preserved.

In case of `// startIndex is a cool argument` comment, the replacement cannot be done using the analyzer unless you search separately for the comment nodes and you have no guarantees the string 'startIndex' in the comment really matches with the token in the source code.

**Comment by KlausLoeffelmann** [comment]
Rafaelle, please, give me just a chance to ask this question:
Let's assume for a moment, we cannot compile AFTER the fix had been applied.

(I would not go on your nerves like this, if not 190 (!) files would be affected, and I am sorry for being so adamant. When it turns out I have been unjustified concerned all along, I promise to eat humble pie!!)

Can the Analyzer be sure _**to apply the CodeFix only**_ when it has semantically proofed that the argument to be nameof'ed has really been passed or that, if it is inside a property's getter, that the Argument equals the property's name? And if not, that it does not apply a code fix, so leaving the original line in place?

**Comment by raffaeler** [comment]
@KlausLoeffelmann oh please, the assumption is just wrong. We are in 2018 and CI / gated checkin is a must and is currently in place on this repo.
So please, let's not do philosophy.
Sorry, if you can't/want't reject the PR, let's wait for someone else but it is becoming a huge waste of time.

**Comment by AdamYoblick** [comment]
Hi @raffaeler, we're going to discuss this internally. I'm not sure if the contribution guidelines mention running an analyzer for PR's. 

If we decide this is something we want to happen for all PRs, perhaps we should be running it automatically as part of our CI build or a git trigger instead of having contributors run it.

Going open is new to our team so we will have lots of wrinkles to iron out. Please be patient while we figure out what we want to do. Thanks :)

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 2]
```diff
@@ -424,6 +424,23 @@ public virtual void SetAudio(byte[] audioBytes)
...
+        /// <devdoc>
+        ///    <para>[To be supplied.]</para>
+        /// </devdoc>
+        public virtual void SetAudio(Span<byte> audioBytes)
```
Comment (karelz) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: Why is this new property part of the change?

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 3]
```diff
@@ -1034,10 +1051,11 @@ private int SaveStreamToHandle(ref IntPtr handle, Stream stream)
             }
             try
             {
-                byte[] bytes = new byte[size];
-                stream.Position = 0;
-                stream.Read(bytes, 0, size);
-                Marshal.Copy(bytes, 0, ptr, size);
+                unsafe
```
Comment (karelz) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: This should be separate change

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 4]
```diff
@@ -1972,9 +1990,11 @@ private unsafe string ReadHtmlFromHandle(IntPtr handle)
                 try
                 {
                     int size = UnsafeNativeMethods.GlobalSize(new HandleRef(null, handle));
-                    byte[] bytes = new byte[size];
-                    Marshal.Copy(ptr, bytes, 0, size);
-                    stringData = Encoding.UTF8.GetString(bytes);
+                    unsafe
```
Comment (karelz) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: This as well

File: src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs [hunk 5]
```diff
@@ -2,56 +2,70 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-namespace System.Windows.Forms {
+namespace System.Windows.Forms
```
Comment (karelz) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs]: Changes in this file should be in separate PR

**Comment by raffaeler** [comment]
@AdamYoblick thank you for the comment.
I remember to have carefully read the guidelines when I first submitted the PR (the one that was accidentally closed) when the repo was still private. I don't remember any mention to analyzers.
Anyway I understand the problem and the decision is yours.

In the past I had to make similar changes in a large codebase and, even if I already wrote an analyzer, I had to change strategy for the reason I mentioned.
All the best.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs (hunk 2 above)
Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: This is my fault when I created this new PR. The one that was accidentally closed did not contain that, but I clearly made a mistake in migrating the changes on the repo I had to re-fork because of that.
These changes are already part of another PR, therefore they make no sense here and are 'duplicated'
I will now restore the code. Sorry.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs (hunk 3 above)
Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: sure, same reason

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs (hunk 4 above)
Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs]: idem

File: src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs (hunk 5 above)
Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs]: indeed

**Comment by karelz** [comment]
BTW I believe @dotMorten has some analyzer doing similar things (for future use in other repos, etc.).

> @raffaeler @MarcoRossignoli a gist for the regex? I understand they need to verify the results one by one in any case.
> I wrote and use a lot of Roslyn based code, but more the analyzer is complex, more difficult is validating the analyzer code, blindly trusting the results.

Agreed. The change should be ideally verified one by one. I would not trust a regex unless I created it myself 😄 ... and even then ...
It would be different story if it was review taking hours. I would change my view point in such case. This one took ~15 mins, so ok-ish (for me).

For reference: For larger changes generated by tools/regex/etc., it is reasonable to ask for the source code of the tool, then verify it and run the same tool locally to verify the results match with the PR. Each person has different threshold when this triggers as a "must have", which is fine -- we are all different individuals with different preferences after all.

> @raffaeler Sorry if the PR is large, but creating multiple PRs does not make that easier.

Agreed.
In future, the only minor improvement might be to separate out the "real" changes where the name was incorrect into separate PR. In this case, it is just 9 cases and makes reviewer alert 😉, so just nice to have IMO.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs (hunk 5 above)
Comment (raffaeler) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataStreamFromComStream.cs]: I now restored the original file. Is it possible that VS automatically formatted the code?

**Comment by karelz** [comment]
Thank you! Reviewed and merged!

**Comment by raffaeler** [comment]
Thank you!! Harder then expected :-)

**Comment by karelz** [comment]
Isn't harder than expected the definition of SW engineer's daily life? 😉
I am almost daily surprised that things/approaches I think are obvious are viewed entirely differently by others. SW engineers thrive on disagreements and discussions (incl. myself) 😄.

**Comment by raffaeler** [comment]
@karelz LOL, I can't live a single day of my life writing code even if most of the time I am involved in architectural roles (which I love too).
Until the discussion is just because of different technical perspectives it is always good even if there is disagreement.
Then we know that different cultures have different reactions. In this case travelling is the only treatment :)
